Add Temporal support

rkh opened the pull request:
This is an alternative to #635, adding only support for Temporal time objects without implementing a public variable caster API.

It still has some special handling to work fine in environments without Temporal. It also avoids running any such checks at load time, to allow loading a Temporal polyfill after loading fluent-bundle.

## Duration

It does not implement `Temporal.Duration` support yet, which would require:
* Augmenting built-in TypeScript `Intl` namespace to support `Intl.DurationFormat` and `Intl.DurationFormatOptions`
* Add signature for `Intl.DurationFormat` to `scope.memoizeIntlObject`
* Add `FluentDuration` wrapper for `Temporal.Duration`
* Optional: Add `DURATION` builtin

Happy to work on that as well, but can also create a separate PR for that.

## FluentVariable

I've taken the liberty to move `FluentVariable` to `types.js` – that made more sense to me, especially with the expanded list.

eemeli reviewed (changes requested):
Looks very promising!

rkh commented on fluent-bundle/src/temporal.d.ts:
I do think `import type { Temporal } from "temporal-polyfill"` would be the most ideal, assuming their type definitions mirror the correct ones, and I think the built result would only impact type definitions.

For our app, I'm just straight up declaring the Temporal namespace, which makes this very easy, but that only works in build environments that don't support Temporal (compiled result works fine in those that do).

I was trying that initially, but ran into issues with using these as `instanceof` values, and `(globalThis as any).Temporal as Temporal` didn't work either. My first attempt ended up with quite a bit of `@ts-ignore` in there.

Can look into this further.

rkh replied:
I have done this for now. However, this will make `temporal-polyfill` a dev dependency for anyone using `fluent-bundle` with TypeScript (see types.d.ts).

Could improve the situation by adding `temporal-spec` as an actual dependency or stating so in the readme.

Demivan commented on fluent-bundle/test/temporal_test.js:
```diff
@@ -0,0 +1,186 @@
+  }
+
+  suiteSetup(async function () {
+    if (typeof Temporal === "undefined") {
+      await import("temporal-polyfill/global");
+
+      bundle = new FluentBundle("en-US", { useIsolating: false });
+      bundle.addResource(
+        new FluentResource(ftl`
+        direct = { $arg }
+        dt = { DATETIME($arg) }
+        month = { DATETIME($arg, month: "long", year: "numeric") }
+        timezone = { DATETIME($arg, timeZoneName: "shortGeneric") }
+        `)
+      );
+    }
```
Setting up the bundle should not be done conditionally, just the pollyfilling.

rkh commented on fluent-bundle/src/types.ts:
```diff
@@ -1,12 +1,34 @@
+function isTemporal(value: any, type: "PlainTime"): value is FluentTemporal.PlainTime;
+function isTemporal(value: any, type: "PlainYearMonth"): value is FluentTemporal.PlainYearMonth;
+function isTemporal(value: any, type: "ZonedDateTime"): value is FluentTemporal.ZonedDateTime;
+function isTemporal(value: any, type: keyof FluentTemporal.TemporalContainer): boolean {
+  const temporal = (globalThis as any).Temporal as FluentTemporal.TemporalContainer | undefined;
+  const prototype = temporal && temporal[type];
+  if (!prototype) {
+    return false;
+  }
+  return value instanceof prototype;
+}
```
Ah yes, a Proxy could also solve deferring Temporal checks.

rkh replied:
Solved it without the Proxy (the tests already demonstrate that it is better to resolve this at run time not at load time). Significantly reduced type checking overall with the latest adjustments.

rkh commented on fluent-bundle/src/types.ts:
```diff
@@ -109,32 +131,128 @@ export class FluentNumber extends FluentType<number> {
+  /**
+   * Converts various types representing dates and times to a `FluentDateTime`.
+   */
+  static from(
+    info: FluentDateInfo | FluentType<FluentDateInfo>,
+    opts: Intl.DateTimeFormatOptions = {}
+  ): FluentDateTime {
```
Good point. I was mirroring what Temporal itself does.

rkh commented on fluent-bundle/src/types.ts:
```diff
@@ -1,12 +1,34 @@
+export type FluentStrictDateValue = FluentTemporal.Plain | Date;
+export type FluentDateValue = FluentStrictDateValue | number;
+export type FluentDateInfo = FluentDateValue | Date | FluentTemporal.Instant | FluentTemporal.ZonedDateTime;
+export type FluentVariable = FluentValue | FluentDateInfo | string | number;
```
Removed and inlined.

rkh commented:
Incorporated the feedback and added more tests (different calendars, number conversion).

The main thing I see right now is the type import, which creates a TypeScript-only dependency on temporal-polyfill.

eemeli commented on fluent-bundle/src/types.ts:
```diff
@@ -1,7 +1,20 @@
+  | Temporal.PlainMonthDay
+  | Temporal.ZonedDateTime
+  | string
+  | number;
```
```suggestion
  | number
  | Date;
```

eemeli commented on fluent-bundle/src/types.ts:
```diff
@@ -104,37 +117,134 @@ export class FluentNumber extends FluentType<number> {
+    // Temporary workaround to support environments without Temporal
+    if ('Temporal' in globalThis) {
+      if (
+        // @ts-ignore
+        value instanceof Temporal.Instant        || // @ts-ignore
+        value instanceof Temporal.PlainDateTime  || // @ts-ignore
+        value instanceof Temporal.PlainDate      || // @ts-ignore
+        value instanceof Temporal.PlainMonthDay  || // @ts-ignore
+        value instanceof Temporal.PlainTime      || // @ts-ignore
+        value instanceof Temporal.PlainYearMonth || // @ts-ignore
+        value instanceof Temporal.ZonedDateTime
```
Why are all these ts-ignore directives required? If they're needed, could ts-expect-error be used instead?

rkh replied:
They are needed because `Temporal` is only imported as a type, but not declared. Will look into `ts-expect-error`.

eemeli commented on fluent-bundle/src/types.ts:
```diff
@@ -104,37 +117,134 @@ export class FluentNumber extends FluentType<number> {
   /** Options passed to `Intl.DateTimeFormat`. */
   public opts: Intl.DateTimeFormatOptions;
 
+  static supportsValue(value: any): value is ConstructorParameters<typeof Temporal.Instant>[0] {
```
Isn't `ConstructorParameters<typeof Temporal.Instant>[0]` the `epochNanoseconds` bigint? That seems wrong here.

rkh replied:
Yes, that is wrong, not sure how that happened. I meant it to be `typeof FluentDateTime`.

eemeli commented on fluent-bundle/src/types.ts:
```diff
@@ -104,37 +117,134 @@ export class FluentNumber extends FluentType<number> {
+    }
+
+    if (typeof value === "object") {
+      // Intl.DateTimeFormat defaults to gregorian calendar, but Temporal defaults to iso8601
+      if ('calendarId' in value) {
+        if (opts.calendar === undefined) {
+          opts = { ...opts, calendar: value.calendarId };
+        } else if (opts.calendar !== value.calendarId && 'withCalendar' in value) {
+          value = value.withCalendar(opts.calendar);
+        }
+      }
```
This automatic conversion is probably a bad idea, esp. as the formatter option `calendar` is already excluded from `DATETIME_ALLOWED`.

rkh replied:
Okay, makes sense. But setting the option is acceptable? I wanted to avoid users having to create their own `FluentDateTime`, and as I said in the comment, `Temporal` defaults to a different calendar than `Intl`.

eemeli replied:
Yeah, if no `calendar` value is set, then it's reasonable to default to the one in the temporal instance.

rkh replied:
That sounds good (and reasonable). I do wish that this was the `Intl.DateTimeFormat` behavior to begin with.

eemeli commented on fluent-bundle/src/types.ts:
```diff
@@ -104,37 +117,134 @@ export class FluentNumber extends FluentType<number> {
+        }
+      }
+
+      // Temporal.ZonedDateTime is timezone aware
+      if ('timeZoneId' in value) {
+        if (opts.timeZone === undefined) {
+          opts = { ...opts, timeZone: value.timeZoneId };
+        } else if (opts.timeZone !== value.timeZoneId && 'withTimeZone' in value) {
+          value = value.withTimeZone(opts.timeZone);
+        }
+      }
+
+      // Temporal.ZonedDateTime cannot be formatted directly
+      if ('toInstant' in value) {
+        value = value.toInstant();
+      }
```
This automatic conversion is also a bad idea. As your comment notes, zoned datetimes are not formattable at least for now, so let's stick to that in Fluent as well.

eemeli reviewed (changes requested):
I'm a bit concerned about the temporal types getting included in the FluentVariable and FluentDateTime external types, as that's effectively making the `temporal-spec` package a runtime dependency. Could that be avoided somehow, even at the cost of making the types less exact?

rkh commented:
> I'm a bit concerned about the temporal types getting included in the FluentVariable and FluentDateTime external types, as that's effectively making the `temporal-spec` package a runtime dependency. Could that be avoided somehow, even at the cost of making the types less exact?

Yes, that's what I meant with my comment as well. Could type them as Temporal-like objects (which is a bit what I was trying to do in the first version). It would make more sense with what the code is like now, as type checks are only done in a single place. I'll see what I can do without copying over the full type spec.

rkh commented:
FYI, I plan to work on this again later in the week.

rkh commented:
Defined a minimal interface for the few fields the code uses from Temporal. The main issue I can see with that one is that TypeScript at least won't catch if an object is passed in that matches the interface but isn't a Temporal object.

But that's probably preferable over a dependency or shipping types for all the various Temporal classes.

This can be dropped as soon as TypeScript comes with Temporal support.

eemeli reviewed (approved):
This looks good! Sorry for taking so long to reply; I've been on vacation. Once the tests pass, this is good to merge.

rkh commented:
All good. Hope you had a nice vacation.
We've solved this differently for our app weeks ago, so this isn't a blocker for us.
Tests should be passing now.

rkh commented:
Can reproduce the CI failure locally (seems to be Node 18 specific, I wonder if the polyfill is broken on that version).

rkh commented:
I did a one line check whether iso8601 is supported or not, and skip the tests if it isn't. Probably out of scope for fluent.js to have a workaround.
